README.md
BaseMUD
==============
**A thorough code review/overhaul of QuickMUD / ROM 2.4b6 for starting your MUD**

BaseMUD is a work-in-progress fork of QuickMUD that has been heavily
reorganized, refactored, and cleaned-up for clarity and easy extendability. This
project has two main goals:

1. Provide an easy-to-modify codebase for those who want to play around with
  their own stock ROM mud, and
2. Provide a transitional project for porting the ROM codebase to C++ or another
  high-level language.

## Instructions

To compile and run:

```
cd src
make
cd ..
./run.sh
```

OR using Docker to startup and shutdown:

```
docker-compose up

docker-compose down
```

This will mount the player, log, and json/area directories for easier editing of player and area files (when using OLC in-game).

To play:

```
telnet localhost 4000
```

BaseMUD comes packaged with a level 60 "Implementor" account.
The name is `Admin` and the password is `Admin`. *Change the password* and,
if you'd like, create your own god character and raise them to level 60.

## Legal Stuff

BaseMUD is based on [QuickMUD](https://github.com/avinson/rom24-quickmud),
which is based on the ROM 2.4 beta version of Merc 2.1 base code, which is in
turn based on DikuMUD. Credits in licenses:

```
ROM 2.4 copyright (c) 1993-1995 Russ Taylor.
MERC 2.1 code by Hatchet, Furey, and Kahn.
DikuMUD by Hans Staerfeldt, Katja Nyboe, Tom Madsen, Michael Seifert, and Sebastian Hammer
```

If you wish to use BaseMUD, you must comply with the following licenses:

* `doc/license.doc`: DikuMUD license
* `doc/license.txt`: Merc Release 2.1 license
* `doc/rom.license`: ROM 2.4 license

As with QuickMUD, BaseMUD is packaged with the following code (which has been
integrated into the source code more thoroughly):

* OLC 1.81
* Lope's Color 2.0
* Erwin's Copyover
* Erwin's Noteboard
* Color Login

## About this project

This project is still very rough. I've been testing things as I go, but
I've been haphazardly moving large swaths of code around under the premise that
everything will be thoroughly tested later. Hopefully the resulting code will
be modular enough that some sort of unit testing environment will be possible.

Despite the heavy refactoring,
BaseMUD strives to be a pure version of ROM 2.4. However, in the process of sifting
through all this code, sometimes I just can't help myself.
Some small changes and additions have been made, which can be enabled/disabled by
uncommenting/commenting the `#define BASEMUD_(FEATURE)` lines in `src/basemud.h`.
Changes that I've considered bugfixes, typo / spelling error fixes, or patches
to small oversights (i.e, fixing spells that have no casting message) will be
left in.

Most of the code has been manually linted for better readability and brevity.
Some changes are subject to taste, but much care has been spent to make the
code consistent and easy to read. Here are some general style changes:

* Vertical space is prioritized highly - two-line if() statements are bountiful.
* Horizontal space is also cherished. Some lines exceed the old standard 80
  character limit _just barely_ to avoid some ugliness.
* Several giant tables, such as spells, races, and classes, go WAY over the line so
  they can be read and modified as tabular data.
* Redundant comments have been removed for the most part in favor of renaming
  functions, parameters, and variables to be more self-evident.
  The author comments have been preserved out of respect and to comply with licenses.
* _All_ tabs have been removed and replaced with spaces. Everywhere.

Some changes may also be _temporarilly regressive_ as a staging step for some
larger refactoring.  In other words, some things are a mess because I'm still cleaning it :)
The affect code, for example, had all of its easy-to-read assignments replaced with
incomprehensible one-liner functions. The long-term plan, in this case, is to move all of the
affects to a table, at which point this will be fixed.

Note: **OLC and Copyover code is also largely untested, and may even crash! D:**  
  
**Want to contribute or just chat about the project?**  
You can locate us at the [Mud Coder's Guild Slack](https://mudcoders.slack.com); reach out to either Synival or Nikhil for an invite to the private BaseMUD channel.

## Game Changes:

Some small changes have been made, most of which are small quality-of-life
improvements, some of which are personal preferences that will eventually be
phased out or at least optional. A handful of changes were just for fun.

**Major features:**

* The world can now be exported and imported in JSON format.

**Deployment:**

* The `rom` executable is now in the `/bin` directory and run from the _root_
  directory. For convenience, you can use `run.sh` in the root directory.
* The `Dockerfile` has been temporarilly removed until some of the structural
  changes are accounted for.

**Optional Features - enable/disabled by uncommenting/commenting `#define BASEMUD_(FEATURE)` in `basemud.h`:**

* `BASEMUD_SHOW_DOORS`:
    Open and closed doors are now visible in auto exits. This reveals some otherwise
    hidden exits but makes exploring _sooo_ much nicer.
* `BASEMUD_SHOW_ARRIVAL_DIRECTIONS`:
    When a character arrives in the room via standard movement, you will now see in what
    direction they came from.
* `BASEMUD_SHOW_OLC_IN_PROMPT`:
    Added OLC editor and vnum to the front of the prompt by default.
* `BASEMUD_SHOW_ABSOLUTE_HIT_DAMAGE`:
    Hits and misses have additional adjectives to show their power in addition
    to the verb that shows percentage of max hit points. Example: "You wound a scary
    beholder with your **excellent** slash."
* `BASEMUD_SHOW_ENHANCED_DAMAGE`:
    Attacks from successful enhanced damage rolls are now displayed as "heavy <type>".
    Example: "You scratch a lich with your epic **heavy** pierce."
* `BASEMUD_SHOW_RECOVERY_RATE`:
    The `score` command now shows your current hit/mana/move rates. This was a
    debug feature that I felt was nice enough to have.
* `BASEMUD_COLOR_STATUS_EFFECTS`:
    The (Glowing), (White Aura), (Red Aura) etc. tags have appropriate colors.
* `BASEMUD_COLOR_ROOMS_BY_SECTOR`:
    Room titles are colored based on their terrain type.
* `BASEMUD_MORE_PRECISE_CONDITIONS`:
    More status notifications in addition to 'excellent', 'nasty wounds', etc.
    There are now different messages for every interval of 10% of hp.
* `BASEMUD_MORE_PRECISE_RELATIVE_DAMAGE`:
    More low-level damage verbs during combat. The 1~15% damage range verbs
    change from "scratch, graze" to "annoy, scratch, graze, scrape, bruise".
* `BASEMUD_MOBS_SAY_SPELLS`:
    Caster mobs now say their spells when casting just like players.
* `BASEMUD_NO_RECALL_TO_SAME_ROOM`:
    Recall attempts to the target recall room are now ignored completely.
* `BASEMUD_NO_WORTHLESS_SACRIFICES`:
    You can no longer sacrifice worthless objects.
* `BASEMUD_GRADUAL_RECOVERY`:
    Hit / mana / move generation now happens on every pulse rather than in
    singular giant ticks. This is undoubtedly a CPU hog, and will need to be
    improved in some way! In the meantime, it's very luxurious :)
* `BASEMUD_CAP_JOINED_AFFECTS`:
    When affects are joined, the duration will stack as before, but the modifier will
    cap out at the min/max between the two affects. For example: the 'chill touch'
    spell will reduce strength, but when stacked, the strength penalty will use the
    stronger penalty between the old and new affects, rather than add them together
    as before.
* `BASEMUD_ALLOW_STUNNED_MOBS`:
    Mobs can now be in stunned, incapacitated, and mortally wounded states.
* `BASEMUD_DETECT_EXTREME_ALIGNMENTS`:
    Added (Golden Aura) and (Black Aura) for extreme good/evil characters.
* `BASEMUD_MATERIALS_COMMAND`:
    Object / character materials can now be shown with the 'materials' command.
* `BASEMUD_DISENGAGE_COMMAND`:
    Added the `disengage` command to let spellcasters stop fighting and recover mana.
    Doesn't work if they're being targetted by anyone in the room.
* `BASEMUD_ORDER_ALL_COMMAND`:
    Added `@` command as a shortcut for `order all`.
* `BASEMUD_ABILITIES_COMMAND`:
    Added `abilities` command, which shows both skills _and_ spells.
* `BASEMUD_PIXIE_RACE`:
    The "pixie" race has been added just for fun. They can fly, and they're
    small, they detect good/evil, and are extremely frail!
* `BASEMUD_IMPORT_JSON`:
    Imports all `.json` files in the `json/` folder before loading the
    traditional `.are` files.

**Quality-of-Life Features:**

* Current position can be shown in prompt.
* Exits in the prompt now show open/closed status (requires `#define BASEMUD_FEATURES`).
* Stunned / incapacitated / mortally wounded / dead state checks are now based on
  percentage of max hit points.
* The 'lore' command has been implemented. It works like the 'identify' spell,
  but hand-picks information based on your skill percentage. What the player
  knows about is random, seeded by the player's name and the object's vnum.
  It currently cannot be practiced beyond 75%.
* Some, but not all, simple "Ok." messages have been replaced with useful things.
* Lots and lots of warning messages for mismatched doors or keys when loading zones.
  This produces some warning messages from the stock zones (which are warranted).
* Added warnings for `'E'` resets with unequippable wear locations for the item
  or wear locations that are already taken on the mob. Many of these errors are
  acceptable with the stock world, but many indicate real bugs.
* New messages for passing through doors (for both sides).
* Drunkeness, thirst, hunger, and fullness now gives you more useful
  messages when the state changes via drink/eat.
* If standing on/at/in something, The `stand` command without any argument will
  now step off/out.
* Modified the `dump` command to allow `dump stats` or
    `dump world <raw | json>`. Files are dumped to the `dump/` directory.
* Areas already loaded by JSON importing are now ignored.
* Warnings are now produced if non-doors have keys.
* Warnings are now produced for keys that don't exist.

**Gameplay Changes:**

* Door open(always) and closed(w/ `BASEMUD_FEATURES`) statuses are visible in the "exits" command.
* Beds, tents, stools, etc. have automatic flags for `STAND_IN`, `STAND_AT`, etc.
  in addition to some standard hit and move regeneration bonuses.
* Bashing / tripping was effectively useless because combatants were
  automatically moved back to standing position in `violence_update()`. This
  is now a bit smarter, with cooldown rates.
* You cannot change position while in a daze / cooldown state. You can still
  issue several commands while knocked down, but you can't get back up!
* Standing up in combat is automatic once the cooldown period has been reached.
  This is essentially the same as before where you would _always_ stand up during
  combat when attacked, but now it's no longer instant.
* Form / parts for races have been reviewed and tweaked slightly (consistency between
  canids). It makes no gameplay difference, but may in the future.
* Tiny balance changes have been made to certain skills and stat regen amounts.
* Fast healing is only active when sitting, resting, or sleeping.
* Meditation is only active when sitting or resting.

**Patched Oversights:**

* `scan <dir>` only checks two rooms ahead rather than four. To see
  that far, you'll need to use the "farsight" spell :)
* Tweaked some stat tables very slightly to allow for steady progression in
  areas that looked like oversights. (Maximum weight capacity did not increase
  steadily where it easily could have.)
* You can no longer trip targets without feet or legs.
* Looking for doors by name now works the same way as looking for anything else.

CHANGELOG.md
# Changelog for BaseMUD

## Version 0.1.0 Beta 2 (2020-03-14)

**New Commands:**

- Added `jsave` to `wiz_ml.c`: This saves files in the JSON directory. Enter  
    `jsave` without any arguments for more details.

**New Major Features:**

- Added `json_std/` folder for reference only. This is _not_ used by the game,  
    but is rather for administrators that want to disable  
    `BASEMUD_WRITE_EXTENDED_JSON`. In this case, replace `json/` with this  
    directory.

**Bugfixes:**

- Areas from JSON now update at boot time. Whoops!!
- `had_get_by_name()` and `had_get_by_name_exact()` now reference `name`
    instead of `filename`.
- Refreshed `json/` directory, which added some missing flags.

## Version 0.1.0 Beta 1 (2020-03-13)

Special thanks to contributions from:
- [blakepell](https://github.com/blakepell)
- [rhoxie21](https://github.com/rhoxie21)

### SIGNIFICANT CHANGES:

**New Commands:**
* Added `confiscate` command to `wiz_l7.c`. Thanks, [blakepell](https://github.com/blakepell)!
* Added `portals` command. This displays all "portal exits" (labels attached
    to rooms or exits) or "portal links" (connectors between portal exits)
    for either the area or the world.
* Added `redit_portal()` to modify room portal exits.
* Added "portal" command to `redit_change_exit()` to modify exit portal exits.

**New Major Features:**
* Introduced "extended flags", inspired by the more-than-32-bit flag implementation from
    SMAUG. The following macros are available (most correspond to functions): 
    `EXT_ZERO` (Represents "no flags"), `EXT_IS_ZERO()`, `EXT_IS_NONZERO()`, `EXT_BITS()`, 
    `EXT_IS_SET()`, `EXT_EQUALS()`, `EXT_SET()`, `EXT_SET_MANY()`, `EXT_UNSET()`, `EXT_UNSET_MANY()`, 
    `EXT_TOGGLE()`, `EXT_TOGGLE_MANY()`, `EXT_TO_FLAG_T()`, `EXT_INIT_ZERO`, `EXIT_INIT_BITS()`,
    `EXT_FROM_FLAG_T()`, `EXT_FROM_INIT()`, `EXT_WITH()`, `EXT_WITHOUT()`, `EXT_WITH_MANY()`,
    `EXT_WITHOUT_MANY()`, `EXT_INVERTED()`
* Added `races.json` to `json/config`. This replaces the table of races.
* Added `pc_races.json` to `json/config`. This replaces the table of playable races.
* Added `songs.json` to `json/config`. This replaces `songs.txt`.
* Added JSON files for all other in-game tables that could potentially be loaded dynamically.
    They were written at run-time to the `json/unsupported` directory.
* Added "enhanced" JSON format that allows for human-readable line breaks. Because this is
    non-standard, it can be disabled in `basemud.h`.
* Removed all remaining hungarian notation (YAY)

**New Files:**
* Introduced `items.c`, which contains (nearly) all item-type-specific logic. There are many, many
    new helper functions here that dictate item behavior, most of it migrated from `do_*()` functions.
    This is a staging step for migrating objects to a more object-oriented model with polymorphism.
* Introduced `fread.c` and `fwrite.c` for reading/writing basic data to files. Migrated related
    routines from `olc*`, `db.c` and `save.c`.
* Moved `db_export_json()` to new file, `json_export.c` and did a lot of
    refactoring. Most of the nasty macros are gone - yay!
* Introduced `players.c` and `mobiles.c` to separate player-only and mobile-only methods
    that correspond to CHAR_Ts.
* Extracted many area functions to new file `areas.c`.
* Extracted many reset functions to new file `resets.c`.
* Extracted many extra_descr functions to new file `extra_descrs.c`.
* Extracted some help functions to new file `help.c`.
* Moved `qmconfig_read()` to new file `quickmud.c`.

**(Behavioral) Changes:**
* Modified JSON folder structure a bit - `config` is now loaded first, then `areas` then `help`.
    Flags, types, and (new) extended flags are stored in the `meta` directory. New tables that are
    not (yet) read in `config` are in the `unsupported` directory.
* `wizhelp` now shows the minimum level required for each wiz command.
* Replaced all "Ok." messages with more descriptive messages to the player.
* Pressing Control-C now gracefully triggers a shutdown. Hitting Control-C again will forcefully quit.
* Made several properties in JSON objects optional.
* Removed redundant newlines from JSON strings.
* `abilities` and `spells` commands now shows both mana _and_ percentage learned for spells.

**Major Internal Changes:**
* Completely overhauled the ROM skill group system. "Groups" have been renamed to "skill groups"
    for consistency.
* Replaced _all_ linked lists with double listed lists.
* Added links to parent entities to all child entity structures, including
    entities that can be linked to multiple parents like objects, extra_descrs,
    and affects. This is mostly so entities can detach themselves to their
    parents easilly upon disposal.
* Replaced most linked list modification to their own functions so lists are
    more reliably maintained between various methods of entity loading/unloading.
* Removed `gsn_*` for global skills. Please use `SN(SKILL)` instead (ex: `SN(FRENZY)`).
* Lots and lots and LOTS of rewrites to portals to make them work better with
    OLC. Many dumb things were replaced, many new features were added, and they
    are (at least comparitively) much easier to work with.

### SMALLER CHANGES:

**New Minor Features:**
* Modifying exits via OLC now automatically manages portal exits and portal
    links. Portal exits/links deleted or created are logged to the player.
* Added `min_pos` property to socials in `json/config/socials.json`. This is mainly for the
    `snore` social, which was hard-coded to be allowed during sleeping.
* Replaced player flags and mob flags with new extended flags.
* Added dotted lines in `practice` and `skills` commands between skills and their mana/percent values
    for better readability. This can be disabled via `basemud.h`.

src/boot.c
/***************************************************************************
 *  Original Diku Mud copyright (C) 1990, 1991 by Sebastian Hammer,        *
 *  Michael Seifert, Hans Henrik Strfeldt, Tom Madsen, and Katja Nyboe.    *
 *                                                                         *
 *  Merc Diku Mud improvments copyright (C) 1992, 1993 by Michael          *
 *  Chastain, Michael Quan, and Mitchell Tse.                              *
 *                                                                         *
 *  In order to use any part of this Merc Diku Mud, you must comply with   *
 *  both the original Diku license in 'license.doc' as well the Merc       *
 *  license in 'license.txt'.  In particular, you may not remove either of *
 *  these copyright notices.                                               *
 *                                                                         *
 *  Thanks to abaddon for proof-reading our comm.c and pointing out bugs.  *
 *  Any remaining bugs are, of course, our work, not his.  :)              *
 *                                                                         *
 *  Much time and thought has gone into this software and you are          *
 *  benefitting.  We hope that you share your changes too.  What goes      *
 *  around, comes around.                                                  *
 ***************************************************************************/

/***************************************************************************
 *  ROM 2.4 is copyright 1993-1998 Russ Taylor                             *
 *  ROM has been brought to you by the ROM consortium                      *
 *      Russ Taylor (rtaylor@hypercube.org)                                *
 *      Gabrielle Taylor (gtaylor@hypercube.org)                           *
 *      Brian Moore (zump@rom.org)                                         *
 *  By using this code, you have agreed to follow the terms of the         *
 *  ROM license, in the file Rom24/doc/rom.license                         *
 ***************************************************************************/

#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <errno.h>

#if defined(unix)
    #include <signal.h>
#endif

#include "signal.h"
#include "globals.h"
#include "utils.h"
#include "interp.h"
#include "db.h"
#include "descs.h"
#include "recycle.h"
#include "memory.h"
#include "save.h"
#include "chars.h"
#include "string.h"
#include "olc.h"
#include "nanny.h"
#include "update.h"
#include "comm.h"
#include "rooms.h"
#include "quickmud.h"

#include "act_info.h"

#include "boot.h"

int main (int argc, char **argv) {
    struct timeval now_time;
    bool copyover = FALSE;
    int free_count;
    #ifdef IMC
        int imcsocket = -1;
    #endif

    /* Memory debugging if needed. */
    #if defined(MALLOC_DEBUG)
        malloc_debug (2);
    #endif

    /* Catch CTRL-C so we can do some manditory clean-up. */
    init_signal_handlers ();

    /* Init time.  */
    gettimeofday (&now_time, NULL);
    current_time = (time_t) now_time.tv_sec;
    strcpy (str_boot_time, ctime_fixed (&current_time));

    /* Macintosh console initialization. */
    #if defined(macintosh)
        console_options.nrows = 31;
        cshow (stdout);
        csetmode (C_RAW, stdin);
        cecho2file ("log file", 1, stderr);
    #endif

    /* Reserve one channel for our use. */
    if ((reserve_file = fopen (NULL_FILE, "r")) == NULL) {
        perror (NULL_FILE);
        exit (1);
    }

    /* Get the port number. */
    port = 4000;
    if (argc > 1) {
        if (!is_number (argv[1])) {
            fprintf (stderr, "Usage: %s [port #]\n", argv[0]);
            exit (1);
        }
        else if ((port = atoi (argv[1])) <= 1024) {
            fprintf (stderr, "Port number must be above 1024.\n");
            exit (1);
        }

        /* Are we recovering from a copyover? */
        if (argv[2] && argv[2][0]) {
            copyover = TRUE;
            control = atoi (argv[3]);
            #ifdef IMC
                imcsocket = atoi (argv[4]);
            #endif
        }
        else
            copyover = FALSE;
    }

    /* Run the game. */
    qmconfig_read(); /* Here because it fits, no conflicts with Linux placement -- JR 05/06/01 */
                     /* Here so we can set the IP adress. -- JR 05/06/01 */

    #if defined(macintosh) || defined(MSDOS)
        boot_db ();
        log_f ("ROM is ready to rock on port %d (%s).", port, mud_ipaddress);
        game_loop_mac_msdos ();
    #endif

    #if defined(unix)
        if (!copyover)
            control = init_socket (port);
        boot_db ();
        log_f ("ROM is ready to rock on port %d (%s).", port, mud_ipaddress);

        #ifdef IMC
            /* Initialize and connect to IMC2 */
            imc_startup (FALSE, imcsocket, copyover);
        #endif

        if (copyover)
            copyover_recover ();

        game_loop_unix (control);
        close (control);
        #ifdef IMC
            imc_shutdown (FALSE);
        #endif
    #endif

    /* Free allocated memory so we can track what was lost due to
     * memory leaks. */
    log_string ("Freeing all objects.");
    free_count = recycle_free_all ();
    log_f ("   %d object(s) freed.", free_count);
    log_string ("Freeing all tables.");
    table_dispose_all ();

    /* Free allocated memory so we can track what was lost due to
     * memory leaks. */
#ifndef BASEMUD_DEBUG_DISABLE_MEMORY_MANAGEMENT
    log_string ("Freeing all allocated memory.");
    free_count = string_space_dispose ();
    log_f ("   %d bytes of string space freed.", free_count);
    free_count = mem_pages_dispose ();
    log_f ("   %d bytes of paged memory freed.", free_count);
#endif

    /* Close our reserved file. */
    fclose (reserve_file);

    /* That's all, folks. */
    log_string ("Normal termination of game.");
    exit (0);
    return 0;
}

#if defined(macintosh) || defined(MSDOS)
void game_loop_mac_msdos (void) {
    struct timeval last_time;
    struct timeval now_time;
    static DESCRIPTOR_T dcon;

    gettimeofday (&last_time, NULL);
    current_time = (time_t) last_time.tv_sec;

    /* New_descriptor analogue. */
    dcon.descriptor = 0;
    if (!mud_ansiprompt)
        dcon.connected = CON_GET_NAME;
    else
        dcon.connected = CON_ANSI;

    dcon.ansi          = mud_ansicolor;
    dcon.host          = str_dup ("localhost");
    dcon.outsize       = 2000;
    dcon.outbuf        = mem_alloc (dcon.outsize);
    dcon.next          = descriptor_list;
    dcon.showstr_head  = NULL;
    dcon.showstr_point = NULL;
    dcon.olc_edit      = NULL; /* OLC */
    dcon.string_edit   = NULL; /* OLC */
    dcon.editor        = 0;    /* OLC */
    descriptor_list    = &dcon;

    /* First Contact! */
    if (!mud_ansiprompt) {
        extern char *help_greeting;
        if ( help_greeting[0] == '.' )
            send_to_desc ( help_greeting+1, &dcon );
        else
            send_to_desc ( help_greeting  , &dcon );
    }
    else
        write_to_buffer (&dcon, "Do you want ANSI? (Y/n) ", 0);

    /* Main loop */
    merc_down = FALSE;
    in_game_loop = TRUE;
    while (!merc_down) {
        DESCRIPTOR_T *d;

        /* Process input. */
        for (d = descriptor_list; d != NULL; d = d_next) {
            d_next = d->next;
            d->fcommand = FALSE;

            #if defined(MSDOS)
                if (kbhit ())
            #endif
            {
                if (d->character != NULL)
                    d->character->timer = 0;
                if (!read_from_descriptor (d)) {
                    if (d->character != NULL && d->connected == CON_PLAYING)
                        save_char_obj (d->character);
                    d->outtop = 0;
                    close_socket (d);
                    continue;
                }
            }

            if (d->character != NULL && d->character->wait > 0)
                continue;

            read_from_buffer (d);
            if (d->incomm[0] != '\0') {
                d->fcommand = TRUE;
                char_stop_idling (d->character);

                /* OLC */
                if (d->showstr_point) {
                    d->lines_written = 0;
                    show_page (d);
                }
                else if (d->string_edit)
                    string_add (d->character, d->incomm);
                else {
                    switch (d->connected) {
                        case CON_PLAYING:
                            if (!run_olc_editor (d))
                                desc_substitute_alias (d, d->incomm);
                            break;
                        default:
                            nanny (d, d->incomm);
                            break;
                    }
                }

                d->incomm[0] = '\0';
            }
        }

        /* Autonomous game motion.  */
        update_handler ();

        /* Output. */
        for (d = descriptor_list; d != NULL; d = d_next) {
            d_next = d->next;
            if ((d->fcommand || d->outtop > 0)) {
                if (!process_output (d, TRUE)) {
                    if (d->character != NULL && d->connected == CON_PLAYING)
                        save_char_obj (d->character);
                    d->outtop = 0;
                    close_socket (d);
                }
            }
        }

        /* Synchronize to a clock.
         * Busy wait (blargh). */
        now_time = last_time;
        while (1) {
            int delta;

            #if defined(MSDOS)
                if (kbhit ())
            #endif
            {
                if (dcon.character != NULL)
                    dcon.character->timer = 0;
                if (!read_from_descriptor (&dcon)) {
                    if (dcon.character != NULL && d->connected == CON_PLAYING)
                        save_char_obj (d->character);
                    dcon.outtop = 0;
                    close_socket (&dcon);
                }
                #if defined(MSDOS)
                    break;
                #endif
            }

            gettimeofday (&now_time, NULL);
            delta = (now_time.tv_sec - last_time.tv_sec) * 1000 * 1000
                + (now_time.tv_usec - last_time.tv_usec);
            if (delta >= (1000000 / PULSE_PER_SECOND) * 100 / PULSE_SPEED)
                break;
        }
        last_time = now_time;
        current_time = (time_t) last_time.tv_sec;
    }
    in_game_loop = FALSE;
}
#endif

#if defined(unix)
void game_loop_unix (int control) {
    static struct timeval null_time;
    struct timeval last_time;

    signal (SIGPIPE, SIG_IGN);
    gettimeofday (&last_time, NULL);
    current_time = (time_t) last_time.tv_sec;

    /* Main loop */
    merc_down = FALSE;
    in_game_loop = TRUE;
    while (!merc_down) {
        fd_set in_set;
        fd_set out_set;
        fd_set exc_set;
        DESCRIPTOR_T *d;
        int maxdesc;

#if defined(MALLOC_DEBUG)
        if (malloc_verify () != 1)
            abort ();
#endif

        /* Poll all active descriptors.  */
        FD_ZERO (&in_set);
        FD_ZERO (&out_set);
        FD_ZERO (&exc_set);
        FD_SET (control, &in_set);
        maxdesc = control;
        for (d = descriptor_first; d; d = d->global_next) {
            maxdesc = UMAX (maxdesc, d->descriptor);
            FD_SET (d->descriptor, &in_set);
            FD_SET (d->descriptor, &out_set);
            FD_SET (d->descriptor, &exc_set);
        }

        if (select (maxdesc + 1, &in_set, &out_set, &exc_set, &null_time) < 0) {
            perror ("Game_loop: select: poll");
            exit (1);
        }

        /* New connection? */
        if (FD_ISSET (control, &in_set))
            init_descriptor (control);

        /* Kick out the freaky folks. */
        for (d = descriptor_first; d != NULL; d = d_next) {
            d_next = d->global_next;
            if (FD_ISSET (d->descriptor, &exc_set)) {
                FD_CLR (d->descriptor, &in_set);
                FD_CLR (d->descriptor, &out_set);
                if (d->character && d->connected == CON_PLAYING)
                    save_char_obj (d->character);
                d->outtop = 0;
                close_socket (d);
            }
        }

        /* Process input. */
        for (d = descriptor_first; d != NULL; d = d_next) {
            d_next = d->global_next;
            d->fcommand = FALSE;

            if (FD_ISSET (d->descriptor, &in_set)) {
                if (d->character != NULL)
                    d->character->timer = 0;
                if (!read_from_descriptor (d)) {
                    FD_CLR (d->descriptor, &out_set);
                    if (d->character != NULL && d->connected == CON_PLAYING)
                        save_char_obj (d->character);
                    d->outtop = 0;
                    close_socket (d);
                    continue;
                }
            }

            if (d->character != NULL && d->character->wait > 0)
                continue;

            read_from_buffer (d);
            if (d->incomm[0] != '\0') {
                d->fcommand = TRUE;
                char_stop_idling (d->character);

                /* OLC */
                if (d->showstr_point) {
                    d->lines_written = 0;
                    show_page (d);
                }
                else if (d->string_edit)
                    string_add (d->character, d->incomm);
                else {
                    switch (d->connected) {
                        case CON_PLAYING:
                            if (!run_olc_editor (d))
                                desc_substitute_alias (d, d->incomm);
                            break;
                        default:
                            nanny (d, d->incomm);
                            break;
                    }
                }
                d->incomm[0] = '\0';
            }
        }

        #ifdef IMC
            imc_loop();
        #endif

        /* Autonomous game motion. */
        update_handler ();

        /* Output. */
        for (d = descriptor_first; d != NULL; d = d_next) {
            d_next = d->global_next;

            if ((d->fcommand || d->outtop > 0)
                && FD_ISSET (d->descriptor, &out_set))
            {
                if (!process_output (d, TRUE)) {
                    if (d->character != NULL && d->connected == CON_PLAYING)
                        save_char_obj (d->character);
                    d->outtop = 0;
                    close_socket (d);
                }
            }
        }

        /* Synchronize to a clock.
         * Sleep( last_time + 1/PULSE_PER_SECOND - now ).
         * Careful here of signed versus unsigned arithmetic. */
        {
            struct timeval now_time;
            long secDelta;
            long usecDelta;

            gettimeofday (&now_time, NULL);
            usecDelta = ((int) last_time.tv_usec) - ((int) now_time.tv_usec)
                + ((1000000 / PULSE_PER_SECOND) * 100 / PULSE_SPEED);
            secDelta = ((int) last_time.tv_sec) - ((int) now_time.tv_sec);
            while (usecDelta < 0) {
                usecDelta += 1000000;
                secDelta -= 1;
            }

            while (usecDelta >= 1000000) {
                usecDelta -= 1000000;
                secDelta += 1;
            }

            if (secDelta > 0 || (secDelta == 0 && usecDelta > 0)) {
                struct timeval stall_time;

                stall_time.tv_usec = usecDelta;
                stall_time.tv_sec = secDelta;
                if (select (0, NULL, NULL, NULL, &stall_time) < 0) {
                    if (errno != EINTR) {
                        perror ("Game_loop: select: stall");
                        exit (1);
                    }
                }
            }
        }

        gettimeofday (&last_time, NULL);
        current_time = (time_t) last_time.tv_sec;
    }
    in_game_loop = FALSE;
}
#endif

/* Recover from a copyover - load players */
void copyover_recover (void) {
    DESCRIPTOR_T *d;
    FILE *fp;
    char name[100];
    char host[MSL];
    int desc;
    bool old;

    log_f ("Copyover recovery initiated");
    fp = fopen (COPYOVER_FILE, "r");

    /* there are some descriptors open which will hang forever then ? */
    if (!fp) {
        perror ("copyover_recover:fopen");
        log_f ("Copyover file not found. Exitting.\n\r");
        exit (1);
    }

    /* In case something crashes - doesn't prevent reading  */
    unlink (COPYOVER_FILE);
    while (1) {
        int errorcheck = fscanf (fp, "%d %s %s\n", &desc, name, host);
        if (errorcheck < 0)
            break;
        if (desc == -1)
            break;

        /* Write something, and check if it goes error-free */
        if (!write_to_descriptor (desc, "\n\rRestoring from copyover...\n\r", 0)) {
            close (desc); /* nope */
            continue;
        }

        d = descriptor_new ();
        d->descriptor = desc;

        d->host = str_dup (host);
        LIST2_FRONT (d, global_prev, global_next,
            descriptor_first, descriptor_last);
        d->connected = CON_COPYOVER_RECOVER;    /* -15, so close_socket frees the char */

        /* Now, find the pfile */
        old = load_char_obj (d, name);

        /* Player file not found?! */
        if (!old) {
            write_to_descriptor (desc,
                "\n\rSomehow, your character was lost in the copyover. Sorry.\n\r", 0);
            close_socket (d);
            continue;
        }

        /* Player file found - ok! */
        write_to_descriptor (desc, "\n\rCopyover recovery complete.\n\r", 0);

        /* Just In Case */
        if (!d->character->in_room)
            d->character->in_room = room_get_index (ROOM_VNUM_TEMPLE);

        /* Insert in the char_list */
        LIST2_FRONT (d->character, global_prev, global_next,
            char_first, char_last);

        char_to_room (d->character, d->character->in_room);
        do_look (d->character, "auto");
        act ("$n materializes!", d->character, NULL, NULL, TO_NOTCHAR);
        d->connected = CON_PLAYING;

        if (d->character->pet != NULL) {
            char_to_room (d->character->pet, d->character->in_room);
            act ("$n materializes!.", d->character->pet, NULL, NULL,
                 TO_NOTCHAR);
        }
    }
    fclose (fp);
}
